### PR TITLE
algod: split SetFdSoftLimit calls for relay and non-relay nodes

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

algod sets RLIMIT_NOFILE to max possible value on startup. This might result the node failure on startup especially if it is a non-relaying node and does not need to listen for thousands of incoming connections.

Additionally enhanced config test since I opened the file:
* Replace shared "tempdir" by t.TempDir
* Make them parallel
* Make test naming consistent

## Test Plan

New test for `IsGossipServer` config helper. Existing tests for everything else.

### CODE REVIEW

File: util/util_windows.go [hunk 1]
```diff
@@ -24,6 +24,11 @@ import (
 
 /* misc */
 
+// GetFdSoftLimit returns a current value for file descriptors soft limit.
+func GetFdSoftLimit() (uint64, error) {
+	return 0, nil
```
Comment (AlgoAxel) [inline util/util_windows.go:29]: Is there any consequence for windows node runners here?

**Review by brianolson — DISMISSED** [review]

File: daemon/algod/server.go [hunk 2]
```diff
@@ -109,13 +109,10 @@ func (s *Server) Initialize(cfg config.Local, phonebookAddresses []string, genes
 
 	// Set large enough soft file descriptors limit.
 	var ot basics.OverflowTracker
-	fdRequired := ot.Add(
-		cfg.ReservedFDs,
-		ot.Add(uint64(cfg.IncomingConnectionsLimit), cfg.RestConnectionsHardLimit))
+	fdRequired := ot.Add(cfg.ReservedFDs, cfg.RestConnectionsHardLimit)
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply daemon/algod/server.go:112]: well, the PR title says exactly this

File: util/util_windows.go (hunk 1 above)
Comment (algorandskiy) [reply util/util_windows.go:29]: no, since set is noop

File: config/localTemplate.go [hunk 3]
```diff
@@ -588,3 +588,9 @@ func (cfg Local) TxFilterRawMsgEnabled() bool {
 func (cfg Local) TxFilterCanonicalEnabled() bool {
 	return cfg.TxIncomingFilteringFlags&txFilterCanonical != 0
 }
+
+// IsGossipServer returns true if NetAddress is set and this node supposed
+// to start websocket server
+func (cfg Local) IsGossipServer() bool {
+	return cfg.NetAddress != ""
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply config/localTemplate.go:595]: yes and == check looks suitable there

File: network/wsNetwork.go [hunk 4]
```diff
@@ -724,10 +725,27 @@ func (wn *WebsocketNetwork) setup() {
...
+			if ot.Overflowed {
+				wn.log.Errorf("overflowed when adding up IncomingConnectionsLimit to the existing RLIMIT_NOFILE value; decrease it")
+			} else {
+				err = util.SetFdSoftLimit(fdRequired)
+				if err != nil {
+					wn.log.Errorf("Failed to set a new RLIMIT_NOFILE value to %d: %s", fdRequired, err.Error())
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply network/wsNetwork.go:748]: I opted to have it closer to the place it is used

[the remark this answers is not in the document]
Comment (algorandskiy) [reply network/wsNetwork.go]: moved to server.go

File: config/localTemplate.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (algorandskiy) [reply config/localTemplate.go:595]: I'm not trying to replace that usage. checking NetAddress is natural there

File: daemon/algod/server.go [hunk 5]
```diff
@@ -109,18 +109,27 @@ func (s *Server) Initialize(cfg config.Local, phonebookAddresses []string, genes
...
 		return errors.New(
-			"Initialize() overflowed when adding up ReservedFDs, IncomingConnectionsLimit " +
-				"RestConnectionsHardLimit; decrease them")
+			"Initialize() overflowed when adding up ReservedFDs and RestConnectionsHardLimit; decrease them")
 	}
 	err = util.SetFdSoftLimit(fdRequired)
 	if err != nil {
 		return fmt.Errorf("Initialize() err: %w", err)
 	}
+	if cfg.IsGossipServer() {
```
Comment (bbroder-algo) [inline daemon/algod/server.go:121]: much more clear how it flows with this, thanks.

File: daemon/algod/server.go [hunk 6]
```diff
@@ -109,18 +109,27 @@ func (s *Server) Initialize(cfg config.Local, phonebookAddresses []string, genes
...
+		fdRequired := ot.Add(fdRequired, uint64(cfg.IncomingConnectionsLimit))
+		if ot.Overflowed {
+			return errors.New("Initialize() overflowed when adding up IncomingConnectionsLimit to the existing RLIMIT_NOFILE value; decrease RestConnectionsHardLimit or IncomingConnectionsLimit")
+		}
+		err = util.SetFdSoftLimit(fdRequired)
+		if err != nil {
+			// do not fail but log the error
+			s.log.Errorf("Failed to set a new RLIMIT_NOFILE value to %d: %s", fdRequired, err.Error())
+		}
+	}
```
[the remark this answers is not in the document]
Comment (bbroder-algo) [reply daemon/algod/server.go:132]: maybe 80% of it

[the remark this answers is not in the document]
Comment (algorandskiy) [reply daemon/algod/server.go:143]: added falling back to hard limit if case of failure

[the remark this answers is not in the document]
Comment (algorandskiy) [reply daemon/algod/server.go:141]: I do not like implicitly changing config values (especially this is a copy object). Logging errors should give enough clues to operators

[the remark this answers is not in the document]
Comment (algorandskiy) [reply daemon/algod/server.go:141]: I believe this is unnecessary complication. Moreover, If one opts in to serve REST how can you tell this is less important than serving gossip clients? I completely disagree on trading one limit to another.

[the remark this answers is not in the document]
Comment (algorandskiy) [reply daemon/algod/server.go:141]: and what does it change? the code sets the limit either to `max(ReservedFDs+RestConnectionsHardLimit+IncomingConnectionsLimit, hardFDLimit)`. If REST connections are not used - fine, they yield descriptors to gossip connections. If high number of them is used - then this is configuration problem and we do not know what is the intent of this relay runner.

File: config/localTemplate.go [hunk 7]
```diff
@@ -588,3 +590,38 @@ func (cfg Local) TxFilterRawMsgEnabled() bool {
...
+		cfg.RestConnectionsHardLimit = 2 * reservedRESTConns
+		if cfg.IncomingConnectionsLimit > int(restDelta) {
+			cfg.IncomingConnectionsLimit -= int(restDelta)
+		} else {
+			cfg.IncomingConnectionsLimit = 0
+		}
+	} else {
+		cfg.RestConnectionsHardLimit -= diff
+	}
+	return true
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply config/localTemplate.go]: done

File: util/util_windows.go [hunk 8]
```diff
@@ -24,6 +24,11 @@ import (
 
 /* misc */
 
+// GetFdLimits returns a current values for file descriptors limits.
+func GetFdLimits() (soft uint64, hard uint64, err error) {
+	return 0, 0, nil
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply util/util_windows.go:29]: no, noop is noop. otherwise for Set should be also returning an error

[the remark this answers is not in the document]
Comment (algorandskiy) [reply util/util_windows.go]: opted into returning syscall.RLIM_INFINITY instead

**Review by bbroder-algo — APPROVED** [review]

**Review by gmalouf — APPROVED** [review]

**Pull request merged**